File: java-playground/src/test/java/org/gonza/javaplayground/model/LottoTest.java
package org.gonza.javaplayground.model;

import java.util.List;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class LottoTest {

    @BeforeEach
    void setUp() {
    }

    @Test
    @DisplayName("로또는 6개의 번호로 생성할 수 있다")
    void 정상_생성_테스트() {
        List<LottoNumber> 로또번호들 = List.of(
                new LottoNumber(1),
                new LottoNumber(2),
                new LottoNumber(3),
                new LottoNumber(4),
                new LottoNumber(5),
                new LottoNumber(6)
        );

        Lotto lotto = new Lotto(로또번호들);
        
        List<LottoNumber> 가져온_번호들 = lotto.로또번호_목록_가져오기();
        assertEquals(6, 가져온_번호들.size());
        for (LottoNumber number : 로또번호들) {
            assertTrue(가져온_번호들.contains(number));
        }
    }

    @Test
    @DisplayName("로또 번호가 6개보다 적으면 예외가 발생한다")
    void 적은_번호_예외_테스트() {
        List<LottoNumber> 적은_번호들 = List.of(
                new LottoNumber(1),
                new LottoNumber(2),
                new LottoNumber(3),
                new LottoNumber(4),
                new LottoNumber(5)
        );

        IllegalArgumentException exception = assertThrows(IllegalArgumentException.class,
                () -> new Lotto(적은_번호들));
        assertEquals("로또 번호는 6개여야 합니다.", exception.getMessage());
    }

    @Test
    @DisplayName("로또 번호가 6개보다 많으면 예외가 발생한다")
    void 많은_번호_예외_테스트() {
        List<LottoNumber> 많은_번호들 = List.of(
                new LottoNumber(1),
                new LottoNumber(2),
                new LottoNumber(3),
                new LottoNumber(4),
                new LottoNumber(5),
                new LottoNumber(6),
                new LottoNumber(7)
        );

        IllegalArgumentException exception = assertThrows(IllegalArgumentException.class,
                () -> new Lotto(많은_번호들));
        assertEquals("로또 번호는 6개여야 합니다.", exception.getMessage());
    }

    @Test
    @DisplayName("로또 번호 목록은 불변이다")
    void 불변성_테스트() {
        List<LottoNumber> 번호들 = List.of(
                new LottoNumber(1),
                new LottoNumber(2),
                new LottoNumber(3),
                new LottoNumber(4),
                new LottoNumber(5),
                new LottoNumber(6)
        );
        Lotto lotto = new Lotto(번호들);

        List<LottoNumber> 가져온_번호들 = lotto.로또번호_목록_가져오기();
        assertThrows(UnsupportedOperationException.class, () -> 가져온_번호들.add(new LottoNumber(7)));
    }
}

File: kotlin-playground/src/test/kotlin/org/gonza/kotlinplayground/LottoTest.kt
package org.gonza.kotlinplayground

import org.gonza.kotlinplayground.domain.Lotto
import org.gonza.kotlinplayground.domain.LottoNumber
import org.gonza.kotlinplayground.enum.LottoStatus
import org.gonza.kotlinplayground.utils.LottoConstants
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertDoesNotThrow
import org.junit.jupiter.api.assertThrows
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class LottoTest {
    @Test
    fun `로또는 로또 번호를 지정된 개수만큼 가질 수 있다`() {
        val maxLottoNumberCount = LottoConstants.MAX_LOTTO_NUMBER_HAVE_COUNT
        val invalidLottoNumberList = mutableListOf<LottoNumber>()
        repeat(maxLottoNumberCount + 1) {
            invalidLottoNumberList.add(LottoNumber(number = it + 1))
        }
        val validLottoNumberList = mutableListOf<LottoNumber>()
        repeat(maxLottoNumberCount) {
            validLottoNumberList.add(LottoNumber(number = it + 1))
        }

        assertThrows<IllegalArgumentException> {
            Lotto(lottoNumberList = invalidLottoNumberList, limit = maxLottoNumberCount)
        }
        assertDoesNotThrow {
            Lotto(lottoNumberList = validLottoNumberList, limit = maxLottoNumberCount)
        }
    }

    @Test
    fun `로또의 최초 상태는 UNKNOWN 이다`() {
        val validLottoNumberList = mutableListOf<LottoNumber>()
        repeat(LottoConstants.MAX_LOTTO_NUMBER_HAVE_COUNT) {
            validLottoNumberList.add(LottoNumber(number = it + 1))
        }

        val lotto = Lotto(lottoNumberList = validLottoNumberList)
        val expectedStatus = LottoStatus.ISSUE

        assertEquals(expectedStatus, lotto.status)
    }

    @Test
    fun `로또의 상태가 변경된 경우 다시 상태를 변경할 수 없다`() {
        val validLottoNumberList = mutableListOf<LottoNumber>()
        repeat(LottoConstants.MAX_LOTTO_NUMBER_HAVE_COUNT) {
            validLottoNumberList.add(LottoNumber(number = it + 1))
        }

        val lotto = Lotto(lottoNumberList = validLottoNumberList)
        lotto.updateStatus(LottoStatus.WON)

        assertThrows<IllegalArgumentException> {
            lotto.updateStatus(LottoStatus.ISSUE)
        }
    }

    @Test
    fun `로또 결과를 확인하지 않은 경우 당첨 번호를 알 수 없다`() {
        val validLottoNumberList = mutableListOf<LottoNumber>()
        val answer = mutableListOf<LottoNumber>()
        repeat(LottoConstants.MAX_LOTTO_NUMBER_HAVE_COUNT) {
            validLottoNumberList.add(LottoNumber(number = it + 1))
            answer.add(LottoNumber(number = it + 1))
        }

        val lotto = Lotto(lottoNumberList = validLottoNumberList)

        assertThrows<IllegalArgumentException> { lotto.winnerNumberList }
        assertEquals(LottoStatus.ISSUE, lotto.status)
    }

    @Test
    fun `로또 번호가 3개 이상 일치하는 경우 당첨이 된다`() {
        val validLottoNumberList = mutableListOf<LottoNumber>()
        val answer = mutableListOf<LottoNumber>()
        repeat(LottoConstants.MAX_LOTTO_NUMBER_HAVE_COUNT) {
            validLottoNumberList.add(LottoNumber(number = it + 1))
            answer.add(LottoNumber(number = it + 1))
        }

        val lotto = Lotto(lottoNumberList = validLottoNumberList)

        assertTrue { lotto.compareAll(target = answer) }
        assertTrue { lotto.winnerNumberList.all { it in answer } }
        assertEquals(LottoStatus.WON, lotto.status)
    }

    @Test
    fun `로또 번호가 3개 이상 일치하지 않는 경우 낙첨이 된다`() {
        val validLottoNumberList = mutableListOf<LottoNumber>()
        val answer = mutableListOf<LottoNumber>()
        repeat(LottoConstants.MAX_LOTTO_NUMBER_HAVE_COUNT) {
            validLottoNumberList.add(LottoNumber(number = it + 1))
            answer.add(LottoNumber(number = it + 6))
        }

        val lotto = Lotto(lottoNumberList = validLottoNumberList)

        assertFalse { lotto.compareAll(target = answer) }
        assertTrue { lotto.winnerNumberList.size < LottoConstants.WINNER_CRITERIA }
        assertEquals(LottoStatus.LOST, lotto.status)
    }
}